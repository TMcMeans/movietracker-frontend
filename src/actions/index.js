export const addMovies = movieArray => {
  return {
    type: 'ADD_MOVIES',
    movies: movieArray
  };
};

export const loginUser = user => {
  return {
    type: 'LOGIN_USER',
    user
  };
};

export const logoutUser = () => {
  return {
    type: 'LOGOUT_USER'
  };
};

export const hasErrored = message => ({
  type: 'HAS_ERRORED',
  message
});

export const removeFavorites = () => ({
  type: 'REMOVE_FAVORITES'
});

export const retrieveAllFavorites = favorites => {
  return {
    type: 'RETRIEVE_ALL_FAVORITES',
    favorites
  };
};

export const addMessage = message => {
  return {
    type: 'ADD_MESSAGE',
    message
  };
};
