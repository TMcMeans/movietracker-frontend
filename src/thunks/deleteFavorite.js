import { hasErrored } from '../actions';

export const deleteFavorite = (userId, movieId) => {
  return async (dispatch) => {
    try {
      const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites/${movieId}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json'
        }
      });
      if (!response.ok) {
        throw Error(response.statusText);
      }
      console.log(response)
    } catch(err) {
      dispatch(hasErrored(err.message));
    }
  }
}