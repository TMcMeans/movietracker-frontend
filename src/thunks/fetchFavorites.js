import { hasErrored, retrieveAllFavorites } from "../actions";

export const fetchFavorites = (userId) => {
  return async(dispatch) => {
    try {
      const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites`);
      if (!response.ok) {
        throw Error(response.statusText)
      }
      const result = await response.json()
      console.log(result)
      dispatch(retrieveAllFavorites(result.data))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}