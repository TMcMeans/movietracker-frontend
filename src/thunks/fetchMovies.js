import { addMovies, hasErrored } from '../actions';

export const fetchMovies = url => {
  return async dispatch => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        console.log(response)
        throw Error(response.statusText);
      }
      const movies = await response.json();
      dispatch(addMovies(movies.results));
    } catch (error) {
      dispatch(hasErrored(error.message));
    }
  };
};