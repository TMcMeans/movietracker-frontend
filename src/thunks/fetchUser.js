import { hasErrored, loginUser, addMessage } from "../actions";

export const fetchUser = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetch('http://localhost:3000/api/users', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          email,
          password
        })
      });
      if(!response.ok) {
        dispatch(addMessage('Password or Email does not Match'))
        throw Error(response.statusText)
      }
      const result = await response.json()
      dispatch(loginUser({name: result.data.name, id: result.data.id}))
      dispatch(addMessage('Sucess! You are now Logged in'))
    } catch(err) {
      console.log(err)
      dispatch(hasErrored(err.message))
      dispatch(addMessage('Internal Server Error, Failed to Login'))
    }
  }
}