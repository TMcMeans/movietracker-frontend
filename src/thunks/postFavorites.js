import { hasErrored, addMessage } from '../actions';

export const postFavorites = (movie, userId) => {
  return async dispatch => {
    try {
      const response = await fetch(
        'http://localhost:3000/api/users/favorites/new',
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            ...movie,
            user_id: userId
          })
        }
      );
      if (!response.ok) {
        throw Error(response.statusText);
      }
      const result = await response.json();
      console.log(result)
      dispatch(addMessage('Movie is now being Stalked'));
    } catch (err) {
      dispatch(hasErrored(err.message));
      dispatch(addMessage('Internal Server Error, Failed to Stalk Movie'))
    }
  };
};
