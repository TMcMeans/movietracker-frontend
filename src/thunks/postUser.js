import { hasErrored, loginUser } from "../actions";

export const postUser = (email, password, name) => {
  return async(dispatch) => {
    try {
      const response = await fetch('http://localhost:3000/api/users/new', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          name,
          email,
          password
        })
      });
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const result = await response.json()
      dispatch(loginUser({name, id: result.id}))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}