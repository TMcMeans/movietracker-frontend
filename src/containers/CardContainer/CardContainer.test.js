import {
  CardContainer,
  mapStateToProps,
  mapDispatchToProps
} from './CardContainer';
import { shallow, mount } from 'enzyme';
import React from 'react';
import { postFavorites } from '../../thunks/postFavorites';
import { deleteFavorite } from '../../thunks/deleteFavorite';
import * as actions from '../../actions'; 

describe('CardContainer', () => {
  const mockFunc = jest.fn();
  const match = {
    path: '/favorites'
  };
  const mockMovies = [
    { title: 'Halloween', movie_id: 1},
    { title: 'It', movie_id: 2 },
    { title: 'The Conjuring', movie_id: 3 }
  ]
  const mockFavorites = mockMovies
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <CardContainer
        movies={mockMovies}
        addFavorite={mockFunc}
        favorites={mockFavorites}
        addMessage={mockFunc}
        removeFavorite={mockFunc}
        match={match}
      />
    );
  })
    
  describe('CardContainer Component', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })
  });

  describe('mapStateToProps', () => {
    const mockState = {
      movies: [
        { title: 'Halloween' },
        { title: 'It' },
        { title: 'The Conjuring' }
      ],
      user: { name: 'Tanjie', id: 4 },
      error: '',
      favorites: [
        { title: 'Halloween' },
        { title: 'It' },
        { title: 'The Conjuring' }
      ],
      message: null
    };

    it('should return an array of movies', () => {
      const expected = [
        { title: 'Halloween' },
        { title: 'It' },
        { title: 'The Conjuring' }
      ];

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.movies).toEqual(expected);
    });

    it('should return a user object', () => {
      const expected = { name: 'Tanjie', id: 4 };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.user).toEqual(expected);
    });
    
    it('should return an array of favorites', () => {
      const expected = [
        { title: 'Halloween' },
        { title: 'It' },
        { title: 'The Conjuring' }
      ];
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.favorites).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    
    it('should dispatch postFavorites thunk when addFavorite is called from props', () => {
      const movie = {title: 'Jaws'}
      const thunkToDispatch = postFavorites(movie, 2);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addFavorite(movie, 2);

      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    })

    it('should dispatch deleteFavorite thunk when removeFavorite is called from props', () => {
      const thunkToDispatch = deleteFavorite(2, 3);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.removeFavorite(2, 3);

      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    })

    it('should dispatch addMessage action when addMessage is called from props', () => {
      const actionToDispatch = actions.addMessage('this is my message');

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addMessage('this is my message');

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  })
});
