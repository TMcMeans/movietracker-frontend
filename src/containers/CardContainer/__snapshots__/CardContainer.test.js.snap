// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer CardContainer Component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardContainer
    addFavorite={[MockFunction]}
    addMessage={[MockFunction]}
    favorites={
      Array [
        Object {
          "movie_id": 1,
          "title": "Halloween",
        },
        Object {
          "movie_id": 2,
          "title": "It",
        },
        Object {
          "movie_id": 3,
          "title": "The Conjuring",
        },
      ]
    }
    match={
      Object {
        "path": "/favorites",
      }
    }
    movies={
      Array [
        Object {
          "movie_id": 1,
          "title": "Halloween",
        },
        Object {
          "movie_id": 2,
          "title": "It",
        },
        Object {
          "movie_id": 3,
          "title": "The Conjuring",
        },
      ]
    }
    removeFavorite={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Login to see Stalked Movies",
      "className": "stalked-message",
    },
    "ref": null,
    "rendered": "Login to see Stalked Movies",
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Login to see Stalked Movies",
        "className": "stalked-message",
      },
      "ref": null,
      "rendered": "Login to see Stalked Movies",
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
